pkg/frontend/querymiddleware/running.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+func newFrontendRunningMiddleware(frontendState func() services.State, notRunningBackoffDuration time.Duration, maxRetries int, log log.Logger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return frontendRunningMiddleware{
+			frontendState:             frontendState,
+			notRunningBackoffDuration: notRunningBackoffDuration,
+			maxRetries:                maxRetries,
+			log:                       log,
+
+			next: next,
+		}
+	})
+}
+
+type frontendRunningMiddleware struct {
+	frontendState             func() services.State
+	notRunningBackoffDuration time.Duration
+	maxRetries                int
+	log                       log.Logger
+
+	next Handler
+}
+
+func (f frontendRunningMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	if err := f.waitForRunning(ctx); err != nil {
+		return nil, err
+	}
+
+	return f.next.Do(ctx, r)
+}
+
+func (f frontendRunningMiddleware) waitForRunning(ctx context.Context) (err error) {
+	spanLog, _ := spanlogger.NewWithLogger(ctx, f.log, "waitForRunning")
```

Comment:
we already have a lot of spans and navigating them in tempo is pain. WDYT about adding this span only if there's need for some waiting at all - if wait duration > 0 && current state is not Running.

Reply:
Good point, this is fixed in https://github.com/grafana/mimir/pull/6621/commits/e07d5e9737a4faa0f331860e9462ad4a85c5e9ba as well.